### PR TITLE
arch_atomic: use small lock in libs/libc/machine/arch_atomic.c

### PR DESCRIPTION
**Pull request by hujun260**
## Summary
use small lock in libs/libc/machine/arch_atomic.c

## Impact
atomic

## Testing
ci

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR description does not adequately meet the NuttX requirements.  It's far too brief and lacks crucial information. Here's why and what's missing:

* **Summary:**  "use small lock in libs/libc/machine/arch_atomic.c" is insufficient.  It doesn't explain *why* this change is necessary. Is it a performance optimization? A bug fix?  What problem does it solve?  What specific atomic operations are affected? What is meant by "small lock"? What was used before?

* **Impact:** Simply stating "atomic" doesn't convey enough information.  While it implies changes to atomic operations, it doesn't address the specific impact questions.  Does this change the API?  Does it affect performance (positively or negatively)? Are there any potential downsides?  All the "Impact" checklist items should be explicitly addressed with "YES" or "NO" and a description if "YES".

* **Testing:** "ci" is not acceptable.  The testing section requires details about the *local* testing performed *before* submitting the PR.  What host and target platforms were used?  What specific tests were run?  "ci" suggests reliance on continuous integration, but local verification is essential and must be documented.  Crucially, *actual testing logs* demonstrating the behavior before and after the change are mandatory and missing.


**Example of a better PR description (still needs actual test logs):**

## Summary

This change replaces the existing mutex-based atomic operations in `libs/libc/machine/arch_atomic.c` with a more efficient "small lock" implementation using atomic compare-and-swap operations. This is necessary to improve performance in multi-threaded applications where frequent atomic operations can contend on the mutex, leading to unnecessary overhead. The previous implementation used a full mutex, which is overkill for simple atomic operations. This change specifically affects the `arch_atomic_*` functions.  Fixes #[NuttX Issue Number if applicable].

## Impact

* Is new feature added? NO
* Is existing feature changed? YES.  Atomic operations are now implemented using compare-and-swap.
* Impact on user (will user need to adapt to change)? NO. The API remains the same.
* Impact on build (will build process change)? NO.
* Impact on hardware (will arch(s) / board(s) / driver(s) change)? YES. This change is specific to [list affected architectures].
* Impact on documentation (is update required / provided)? NO.  The API is unchanged.
* Impact on security (any sort of implications)? NO.
* Impact on compatibility (backward/forward/interoperability)? NO.
* Anything else to consider?  Potential performance improvements may vary depending on the target architecture and workload.


## Testing

I confirm that changes are verified on local setup and works as intended:
* Build Host(s): Linux (Ubuntu 20.04), x86_64, GCC 9.4.0
* Target(s): sim:qemu-x86_64


Testing logs before change:

```
[Insert actual logs demonstrating the problem/previous behavior]
```

Testing logs after change:

```
[Insert actual logs demonstrating the improved/corrected behavior]
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by acassis** [comment]
@hujun260 why to use this approach instead of std atomic_load(), similarly what @pussuw did at https://github.com/apache/nuttx/pull/14863 ?

**Comment by xiaoxiang781216** [comment]
> @hujun260 why to use this approach instead of std atomic_load(), similarly what @pussuw did at #14863 ?

@acassis this implement atomic_xxx when compiler doesn't support atomic operation.

**Comment by acassis** [comment]
> > @hujun260 why to use this approach instead of std atomic_load(), similarly what @pussuw did at #14863 ?
> 
> @acassis this implement atomic_xxx when compiler doesn't support atomic operation.

Right, but some that modification from @pussuw will fail if the compiler doesn't support atomic operation, right?

**Comment by acassis** [comment]
@hujun260 could you please review https://nuttx.apache.org/docs/latest/implementation/critical_sections.html to confirm it still correctly after the recent modifications to critical sections and spinlocks?

**Comment by pussuw** [comment]
> > > @hujun260 why to use this approach instead of std atomic_load(), similarly what @pussuw did at #14863 ?
> > 
> > 
> > @acassis this implement atomic_xxx when compiler doesn't support atomic operation.
> 
> Right, but some that modification from @pussuw will fail if the compiler doesn't support atomic operation, right?

It will work even if the compiler does not provide builtin atomics, since then the NuttX library will take over. We will obviously lose any performance benefit in this case, but all targets that support address environments (arm/arm64/risc-v) support compiler atomics as well (at least for word size!). This is why it is very important to select the correctly sized datatype when using these builtin atomics btw.

**Comment by pussuw** [comment]
> > > > > @hujun260 why to use this approach instead of std atomic_load(), similarly what @pussuw did at #14863 ?
> > > > 
> > > > 
> > > > @acassis this implement atomic_xxx when compiler doesn't support atomic operation.
> > > 
> > > 
> > > Right, but some that modification from @pussuw will fail if the compiler doesn't support atomic operation, right?
> > 
> > 
> > It will work even if the compiler does not provide builtin atomics, since then the NuttX library will take over. We will obviously lose any performance benefit in this case, but all targets that support address environments (arm/arm64/risc-v) support compiler atomics as well (at least for word size!). This is why it is very important to select the correctly sized datatype when using these builtin atomics btw.
> 
> So, it could be used here in this PR as well, right?

No, like @xiaoxiang781216 said, if the compiler does not provide builtin atomics like atomic_fetch_add / atomic_fetch_sub and so forth, the NuttX C-library and arch_atomic.c will provide the definitions for them. The selection betweem them happens via https://github.com/apache/nuttx/blob/6944a1630639587ff3eef3ac1499b64811fe8a74/include/nuttx/atomic.h#L77-L94

If the compiler does not provide atomics, nuttx/lib/stdatomic is used instead, it has declarations like this: https://github.com/apache/nuttx/blob/6944a1630639587ff3eef3ac1499b64811fe8a74/include/nuttx/lib/stdatomic.h#L146-L152

And then the definitions for atomic_fetch_add_1 etc. come from arch_atomic.c, notice how they are defined as **weak** symbols, so the compiler's version overrides them.

**Comment by hujun260** [comment]
> @hujun260 could you please review https://nuttx.apache.org/docs/latest/implementation/critical_sections.html to confirm it still correctly after the recent modifications to critical sections and spinlocks?

Yes, this document is still valid.

File: libs/libc/machine/arch_atomic.c [hunk 1]
```diff
@@ -40,23 +46,23 @@
   void weak_function CONCATENATE(fn, n) (FAR volatile void *ptr,   \
                                          type value, int memorder) \
   {                                                                \
-    irqstate_t irqstate = spin_lock_irqsave(NULL);                 \
+    irqstate_t irqstate = spin_lock_irqsave(&g_atomic_lock);       \
```
Comment (anchao) [inline libs/libc/machine/arch_atomic.c:49]: If recursion support is removed, calling `*_wo_note()` version is more safer.

**Comment by anchao** [comment]
and fix typo in commit message:  arc_atomic -> arch_atomic

**Review by xiaoxiang781216 — APPROVED** [review]

File: libs/libc/machine/arch_atomic.c [hunk 2]
```diff
@@ -31,287 +31,293 @@
...
-    if (*tmp == oldvalue)                                        \
-      {                                                          \
-        *tmp = newvalue;                                         \
-      }                                                          \
-                                                                 \
-    spin_unlock_irqrestore(NULL, irqstate);                      \
-    return ret;                                                  \
+#define SYNC_VAL_CMP_SWAP(fn, n, type)                               \
+                                                                     \
+  type weak_function CONCATENATE(fn, n) (FAR volatile void *ptr,     \
```
Comment (xiaoxiang781216) [inline libs/libc/machine/arch_atomic.c:306]: remove ALL space before (

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (hujun260) [reply libs/libc/machine/arch_atomic.c]: ok

**Comment by hujun260** [comment]
> and fix typo in commit message: arc_atomic -> arch_atomic

ok

File: libs/libc/machine/arch_atomic.c (hunk 2 above)
Comment (hujun260) [reply libs/libc/machine/arch_atomic.c]: ok

File: libs/libc/machine/arch_atomic.c [hunk 3]
```diff
@@ -31,287 +31,293 @@
...
-  {                                                                \
-    irqstate_t irqstate = spin_lock_irqsave(NULL);                 \
-                                                                   \
-    *(FAR type *)ptr = value;                                      \
-                                                                   \
-    spin_unlock_irqrestore(NULL, irqstate);                        \
+#define STORE(fn, n, type)                                           \
+                                                                     \
+  void weak_function CONCATENATE(fn, n)(FAR volatile void *ptr,      \
+                                         type value, int memorder)   \
```
Comment (xiaoxiang781216) [inline libs/libc/machine/arch_atomic.c:47]:
```suggestion
                                        type value, int memorder)   \
```
In ALL places

Comment (hujun260) [reply libs/libc/machine/arch_atomic.c]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**